wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Active.hs

```diff
@@ -52,8 +56,36 @@ pay activeWallet pw grouping regulation payment = liftIO $ do
       runExceptT $ do
         (opts, accId, payees) <- withExceptT NewPaymentWalletIdDecodingFailed $
                                    setupPayment policy grouping regulation payment
+
+        -- Verify that all payee addresses are of the same `NetworkMagic`
+        -- as our `ActiveWallet`.
+        let nm = makeNetworkMagic $ Kernel.walletPassive activeWallet ^. walletProtocolMagic
+        verifyPayeesNM nm payees
+
+        -- Pay the payees
         withExceptT NewPaymentError $ ExceptT $
-          Kernel.pay activeWallet pw opts accId payees
+            Kernel.pay activeWallet pw opts accId payees
+
+-- | Verifies that the `NetworkMagic` of each payee address matches the
+-- provided `NetworkMagic`.
+verifyPayeesNM
+    :: Monad m
+    => NetworkMagic
+    -> NonEmpty (Address, Coin)
+    -> ExceptT NewPaymentError m ()
```

Comment:
cf above comment, no need to be unpure here, we can go for `Either SpecificLogicError ()`